nomad/peers/peers.go

```diff
@@ -0,0 +1,265 @@
+			return false, nil
+		}
+	}
+
+	// If the server is missing the rpc_addr tag, default to the serf advertise
+	// addr.
+	rpcIP := net.ParseIP(m.Tags["rpc_addr"])
+	if rpcIP == nil {
+		rpcIP = m.Addr
+	}
+
+	portStr := m.Tags["port"]
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return false, nil
+	}
+
+	buildVersion, err := version.NewVersion(m.Tags["build"])
+	if err != nil {
+		return false, nil
+	}
+
+	raftVsn := 0
+	raftVsnString, ok := m.Tags["raft_vsn"]
+	if ok {
+		raftVsn, err = strconv.Atoi(raftVsnString)
+		if err != nil {
+			return false, nil
+		}
+	}
+
+	// Check if the server is a non voter
+	_, nonVoter := m.Tags["nonvoter"]
+
+	addr := &net.TCPAddr{IP: m.Addr, Port: port}
+	rpcAddr := &net.TCPAddr{IP: rpcIP, Port: port}
+	parts := &Parts{
+		Name:         m.Name,
+		ID:           id,
+		Region:       region,
+		Datacenter:   datacenter,
+		Port:         port,
+		Bootstrap:    bootstrap,
+		Expect:       expect,
+		Addr:         addr,
+		RPCAddr:      rpcAddr,
+		Build:        *buildVersion,
+		RaftVersion:  raftVsn,
+		Status:       m.Status,
+		NonVoter:     nonVoter,
+		MajorVersion: deprecatedAPIMajorVersion,
+	}
+	return true, parts
+}
+
+// PartCache is a threadsafe cache of known Nomad server peers parsed from Serf
+// members. It avoids the need to re-parse Serf members each time the peers
+// need to be inspected.
+type PartCache struct {
+
+	// peers is a map of region names to the list of known server peers in that
+	// region. All access must be protected by peersLock.
+	peers     map[string][]*Parts
+	peersLock sync.RWMutex
+}
+
+// NewPartsCache returns a new instance of a PartsCache ready for use.
+func NewPartsCache() *PartCache {
+	return &PartCache{
+		peers: make(map[string][]*Parts),
+	}
+}
+
+// PeerSet adds or updates the given parts in the cache. This should be called
+// when a new peer is detected or an existing peer changes is status.
+func (p *PartCache) PeerSet(parts *Parts) {
+	p.peersLock.Lock()
+	defer p.peersLock.Unlock()
+
+	existing, ok := p.peers[parts.Region]
+	if !ok {
+		p.peers[parts.Region] = []*Parts{parts}
+		return
+	}
+
+	// Replace if already present
+	for i, ep := range existing {
+		if ep.Name == parts.Name {
+			existing[i] = parts
+			return
+		}
+	}
+
+	// If we reached this point then it's a new member, so append it to the
+	// exiting array.
+	p.peers[parts.Region] = append(existing, parts)
+}
+
+// PeerDelete removes the given members from the cache. This should be called
+// when a peer is reaped from the Serf cluster.
+func (p *PartCache) PeerDelete(event serf.MemberEvent) {
+	p.peersLock.Lock()
+	defer p.peersLock.Unlock()
+
+	for _, m := range event.Members {
+		if ok, parts := IsNomadServer(m); ok {
+			existing := p.peers[parts.Region]
+			n := len(existing)
+			for i := 0; i < n; i++ {
+				if existing[i].Name == parts.Name {
+					existing[i], existing[n-1] = existing[n-1], nil
+					existing = existing[:n-1]
+					n--
+					break
+				}
+			}
```

Comment:
I think this can be replaced by [`slices.DeleteFunc`](https://pkg.go.dev/slices#DeleteFunc) and that'll make sure any potential weird entropy introduced by address changes gets covered. (I suspect serf prevents this internally but I'm not 100% sure we can rely on that with this out-of-library cache.)

Reply:
Updated within https://github.com/hashicorp/nomad/pull/26881/commits/e34355a0f98a0abc07c5c99e6e76ee289f6868af